### ISSUE TITLE
EPI slab to T1 rigid registration fails even after troubleshooting

### ISSUE THREAD

**Issue by ccn30** [ANTsX/ANTs#1110]
**Describe the problem**

Thank you for the wiki/your efforts to make ANTs as user friendly as possible! 
I'm trying to coregister EPI slabs and T1W structurals within subject (EPIs are mean image of first functional run 1.5x1.5x1 voxels and T1W are brain extracted 0.7iso, all data 7T) but 15/18 subjects keep failing (no errors, just non aligned results). I've tried proposed solutions in (slightly) similar issues e.g. #400  or #908 but no success. I'd very grateful for any advice on next steps in troubleshooting. 

I have tried:
- checking the images are read in same orientation in separate ITKSnap windows (they are - I was concerned because the images have been preprocessed with FSL topup/SPM realign reslice for EPIs and SPM BET for T1s so may affect how headers are read)
- comparing images from successful vs unsuccessful regs for inhomogeneities/artefacts but eyeballing they seem similar
- padding the EPIs with 25 voxels to see if boundary was issue, but made it worse in previously successful subjects
- switching T1 to the fixed image and EPI to the moving image instead of vv (some improvements, but many still failed - what is the rationale for warping T1 to EPI in the example #400 ?)
- swtiching from using ANTs/2.1.0 to latest 2.3.4 but no effect

before I spend more time trying different possibilities, do you think:
- I am doing something wrong in the command (e..g shrink factors? i'm fairly new to neuroimaging generally so lots to learn)
- I should try running bias field corrections on the EPIs (there is some inhomogeneity)
- try using EPI as moving image but this time with the padded voxels
- the images are fairly well aligned anyway - is it worth trying to coregister?
- abandon ANTs in this case as the preprocessing from other softwares may be influencing something

Thank you!

**To Reproduce**

I'm calling the command verbatim from [https://github.com/ntustison/PartialSlabEpiT1ImageRegistration/blob/master/commandRigid.sh](url)
${movingImage}=T1.nii
${fixedImage}=epi.nii
${maskImage}=mask.nii
these are attached examples of subject that fails
[epi.nii.tar.gz](https://github.com/ANTsX/ANTs/files/5494408/epi.nii.tar.gz)
[mask.nii.tar.gz](https://github.com/ANTsX/ANTs/files/5494410/mask.nii.tar.gz)
[t1.nii.tar.gz](https://github.com/ANTsX/ANTs/files/5494411/t1.nii.tar.gz)

**System information (please complete the following information)**
Many issues are specific to a particular system. Please include all information
about your computing environment.

 - OS: [e.g. Mac OS] Linux
 - OS version: [e.g. 10.15.1] Debian 8.11 (Jessie)
 - Type of system: [Desktop, laptop, HPC cluster, cloud instance,
   other] HPC cluster

**ANTs version information**
 - ANTs code version: [output of antsRegistration --version] tried both ANTs Version: 2.1.0 and 2.3.4 (however for 2.3.4, --version outputs 0.0.0.0, it was recently compiled by my IT department from GitHub..)
 - ANTs installation type: [Compiled from source, downloaded binary
   (from where?), installed by another package (which?), installed in
   container (URL?), other (please specify)] Compiled from GitHub

**Comment by ntustison** [comment]
Try this modification:

```
#! /bin/sh

baseDirectory=${PWD}

fixedImage=${baseDirectory}/epi.nii
fixedImageN4=${baseDirectory}/epiN4.nii
movingImage=${baseDirectory}/t1.nii
maskImage=${baseDirectory}/mask.nii

# make mask of slab image

N4BiasFieldCorrection -d 3 -i $fixedImage -o $fixedImageN4 -b [100] -c [200x200x200x200,0] -v 1 -s 4

ThresholdImage 3 ${fixedImageN4} ${maskImage} 100 Inf

# Register T1 to EPI slab with mask
outputPrefix=${baseDirectory}/epiSlabxT1

antsRegistration --verbose 1 \
                 --dimensionality 3 \
                 --float 0 \
                 --interpolation Linear \
                 --use-histogram-matching 0 \
                 --winsorize-image-intensities [0.005,0.995] \
                 --output [${outputPrefix},${outputPrefix}Warped.nii.gz,${outputPrefix}InverseWarped.nii.gz] \
                 --initial-moving-transform [${fixedImageN4},${movingImage},1] \
                 --transform translation[0.2] \
                   --metric MI[${fixedImageN4},${movingImage},1,32,Random,0.5] \
                   --convergence [400x200x100,1e-6,10] \
                   --shrink-factors 8x4x2 \
                   --smoothing-sigmas 0x0x0vox \
                   --masks [${maskImageN4},NULL] \
                 --transform Rigid[0.2] \
                   --metric MI[${fixedImageN4},${movingImage},1,32,Random,0.5] \
                   --convergence [500x500x250x50,1e-6,10] \
                   --shrink-factors 6x4x2x1 \
                   --smoothing-sigmas 2x1x0x0vox \
                   --masks [${maskImage},NULL]
```

**Comment by ccn30** [comment]
Thanks for the fast response! Alas these changes seem to make the problem worse - previously successful subjects now fail as well as those that previously failed.
Attached is a screenshot which is representative of most of the registrations fail - the T1W is shifted down and sometimes over rotated. I get similar results using the registration tool in ITKsnap (rigid, MI). Unfortunately I don't have a whole brain EPI to use as an intermediate. 
![Screenshot 2020-11-05 at 19 27 41](https://user-images.githubusercontent.com/47219819/98287785-c75cb400-1f9d-11eb-9b02-bc622e85652f.png)

**Comment by ntustison** [comment]
I don't know how you ran it.  The results I get look pretty good to me.

<img width="1080" alt="Screen Shot 2020-11-05 at 11 42 30 AM" src="https://user-images.githubusercontent.com/324811/98288662-066a1580-1f5c-11eb-962e-3cebc2349fd3.png">

**Comment by ccn30** [comment]
They do look good, thank you! I think it may be a version issue - are you running 2.3.4? I tried to get my IT dept to compile 2.3.4, but I don't think they installed it (antsRegistration --version prints 0.0.0.0 and there are no bin/ or lib/ dirs) so I'm stuck on 2.1.0. I'll get them to fix it and try again.

**Comment by ntustison** [comment]
Built from the latest GitHub repo but I don't know why it would be a version issue as ``antsRegistration`` has been stable for quite some time.  I would make sure that you didn't make a mistake anywhere in running what I wrote above.

**Comment by ccn30** [comment]
You're right, I forgot to change masking threshold down to 100. But didn't solve the issue unfortunately. I copied the above script verbatim, into a new dir with same images I gave you, loaded ANTs/2.3.4, and it still didn't work. Output of command and the warped images attached (can't attach inverse as file too large in GitHub file selector). Does this output look successful on your end? Could it be my ITKsnap?
[epiSlabxT1Warped.nii.gz](https://github.com/ANTsX/ANTs/files/5501773/epiSlabxT1Warped.nii.gz)
[output.txt](https://github.com/ANTsX/ANTs/files/5501774/output.txt)
